src/bloom/hierarchical.rs

```diff
@@ -0,0 +1,335 @@
+            _ => unreachable!(),
+        }
+    }
+
+    fn len(&self) -> usize {
+        match self {
+            Self::Vec(_, values) => values.len(),
+            _ => 0,
+        }
+    }
+
+    fn add_to_intermediate_filters(&mut self, id: usize, item: &[u8]) -> Result<bool> {
+        match self {
+            Self::Vec(filter, values) => {
+                if let Some(filter) = filter {
+                    let mut should_add = false;
+                    for value in values.iter_mut() {
+                        should_add = should_add || value.add_to_intermediate_filters(id, item)?;
+                        if should_add {
+                            filter.add(item)?;
+                        }
+                    }
+                    Ok(should_add)
+                } else {
+                    Ok(false)
+                }
+            }
+            Self::Item(self_id) => Ok(*self_id == id),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<Child, Key> BloomProvider for HierarchicalBloom<Child>
+where
+    Child: BloomProvider<Key = Key> + Send + Sync,
+    Key: Send + Sync + AsRef<[u8]> + ?Sized,
+{
+    type Key = Key;
+
+    async fn check_filter(&self, item: &Self::Key) -> Result<Option<bool>> {
+        self.inner.check_filter(&self.children, item).await
+    }
+
+    async fn offload_buffer(&mut self, needed_memory: usize) -> usize {
+        self.inner
+            .offload_buffer(&mut self.children, needed_memory)
+            .await
+    }
+
+    async fn get_filter(&self) -> Option<Bloom> {
+        match &self.inner {
+            HierarchicalBloomInner::Vec(filter, _) => filter.clone(),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl<Child> HierarchicalBloom<Child>
+where
+    Child: BloomProvider + Send + Sync,
+{
+    /// Create new container
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Count of childs in container
+    pub fn len(&self) -> usize {
+        self.children.len()
+    }
+
+    /// Clear container to default value
+    pub fn clear(&mut self) {
+        *self = Default::default();
+    }
+
+    /// Returns an iterator over the childs
+    pub fn iter(&self) -> std::slice::Iter<'_, Child> {
+        self.children.iter()
+    }
+
+    /// Returns a mutable iterator over the childs
+    pub fn iter_mut(&mut self) -> std::slice::IterMut<'_, Child> {
+        self.children.iter_mut()
+    }
+
+    /// Returns last child
+    pub fn last(&self) -> Option<&Child> {
+        self.children.last()
+    }
+
+    /// Add child to collection
+    pub fn push(&mut self, item: Child) {
+        let id = self.children.len();
+        self.inner.push(id, &None, self.group_size);
+        self.children.push(item);
+    }
+
+    /// Returns children elements as Vec
+    pub fn into_vec(self) -> Vec<Child> {
+        self.children
+    }
+
+    /// Returns mutable reference to inner container. Use carefully
+    pub fn children_mut(&mut self) -> &mut Vec<Child> {
```

Comment:
Exposing inner collection looks unsafe. If it is changed externally, there will be an out of sync between `collection` and `inner`.